### DISCUSSION TITLE
Has anyone successfully ported TinyUSB to work on Zinq7000 processor?

### DISCUSSION THREAD

**Discussion by rjnbarrass** [hathach/tinyusb#1401] [General]
Hello
I'm working with a currently, but possibly FreeRTOS based solution utilising the Zinq7000 processor (development is currently using the Zedboard(s) from Digilent.

We are wishing to use CDC communication for our application but are finding the examples from 2015 are broken so wondered if the TinyUSB stack would be more robust. I unfortunately see that the Zilinx/Zinq processor is not an official supported port, but wondered if anyone had done the work and was willing to freeware that port?

Regards
Richard

**Comment by hathach** [comment]
it is not supported since getting it up and running is not so straightforward. Last time I tried to port it to Pynq 7000 (pink board, gladly donated by one of the user), the vivado software ate all of my ssd spaces. It is just installation, the compilation probably needs even more space.

**Reply by rjnbarrass** [reply]
Thank you - i totally understand the reasons why it might not be a supported platform from your perspective I was just wondering if anyone in the community had done the port and was willing to share  :)
Regards
Richard

**Reply by rjnbarrass** [reply]
Our need for trying to port TinyUSB to work on the Zinq700 / Zedboard/PicoBoard etc has become more of priority. 
So I am attempting to start the porting process... are there any hints and tips on where to start - the device CDC support is my priority initially.
I have both a full 2019.1 and 2021.2 (Windows) and 2021.2 (Unix) build environment available to work with. Plus ideally, integration with FreeRTOS 202012 or later
Regards
Richard

**Reply by tannewt** [reply]
I'm interested in this too. I'm not sure all of the Zynq boards can do device because it requires an external phy. I got a Zybo Z7 because it does support device mode.

My first step would be to get a bare metal example running on the A9. After that I'd check to see if the USB IP is already supported by TinyUSB. (chipidea iirc) I think it might be. If it is, getting a TinyUSB loading would be the next step.

My preference is to get it going without Vivado since its large. I think JLink can load the "FSBL" directly.

**Reply by rjnbarrass** [reply]
Hello
I keep coming to this on/off as other work/project demands allow.
I'm using the Zedboard as my production test board which does have a suitable device USB port (an OTG as it happens)... Using the CDC example code from Xilinx i was able to get the device to enumerate and connect to both WIndows and Linux... however, the sample code was poor and didn't cope with much more than a simple few bytes sent from a terminal to 'prove it works'
As we ideally want to support a double CDC serial channel (and have both reliable) I have turned to TinyUSB.
I have to admit that I am knew to the world of Vivado/Vitis (trying to stick to 2021.2 release toolchain),  but have a good working understanding of those tools and configuring the BSP (including freertos)...

What I am struggling with now is where in the API do I have integrate the following
a) allow TinyUSB to configure the device (I have example code for the various UsbPs calls so just need to know what to map where) -- this would be I assume something that would live under bsp/zinq7000 ? which API's do I need to implement
b) integrate the IRQ call backs to work with FreeRTOS
c) a list of the callbacks to implement to allow sending and receiving data over the enumerated CDC channels.

Regards
Richard

**Reply by hathach** [reply]
@rjnbarrass check out the getting started and porting guide 
- https://github.com/hathach/tinyusb/blob/master/docs/reference/getting_started.rst
- https://github.com/hathach/tinyusb/blob/master/docs/contributing/porting.rst

as mentioned by Scott, Zynq7000 use the same chipidead controller as some of nxp mcu, which is already supported, you may need to add some hal (register read/write, cache clear) https://github.com/hathach/tinyusb/tree/master/src/portable/chipidea/ci_hs for it to work with. Some got it running with host stack though it is a long time ago,

- #62
- #96

**Reply by rjnbarrass** [reply]
Hello
Other project demands have changed and yet again, I return to try and "crack the nut" of getting TinyUSB (device) working on the Zinq7000.
I feel I am a lot further on this time, with integration with the FreeRTOS port as part of the Xilinx BSP set up working too (the interrupt controller definition is hidden inside the FreeRTOS port as powers the 'tick' but when setting up the USB device, you also need to update that same interrupt handler)...
I however think that the defined support from Xilinx (xusbps library) is now confusing or 'limiting' the TinyUSB  implementations.

For info, the main documentation is: [https://xilinx-wiki.atlassian.net/wiki/spaces/A/pages/18842117/Zynq+Standalone+USB+device+driver](url)
API: [https://xilinx.github.io/embeddedsw.github.io/usbps/doc/html/api/index.html](url)

my 'dcd' implementation (still basic, as I debug) is attached
[zinq7000.zip](https://github.com/hathach/tinyusb/files/9288859/zinq7000.zip)

[usbps_v2_7.zip](https://github.com/hathach/tinyusb/files/9288901/usbps_v2_7.zip)

As you can see if dcd_init()

line 145: the USB library interrupt handler is set to call back to my implementation
but
line 154: is the suggested way to configure the interrupt handler for the USB device -- this calls its own handler "XUsbPs_IntrHandler" first - which then calls the handler defined above (and also, any specific end point interrupt handlers)...

Now, I have in usb_descriptors.c the following:

> uint8_t const desc_hs_configuration[] =
{
  // Config number, interface count, string index, total length, attribute, power in mA
  TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, 0x00, 100),
>
> // 1st CDC: Interface number, string index, EP notification address and size, EP data address (out, in) and size.
  TUD_CDC_DESCRIPTOR(ITF_NUM_CDC_0, 4, EPNUM_CDC_0_NOTIF, 8, EPNUM_CDC_0_OUT, EPNUM_CDC_0_IN, 512),
>
>  // 2nd CDC: Interface number, string index, EP notification address and size, EP data address (out, in) and size.
  TUD_CDC_DESCRIPTOR(ITF_NUM_CDC_1, 4, EPNUM_CDC_1_NOTIF, 8, EPNUM_CDC_1_OUT, EPNUM_CDC_1_IN, 512),
};
>
>// device qualifier is mostly similar to device descriptor since we don't change configuration based on speed
tusb_desc_device_qualifier_t const desc_device_qualifier =
{
  .bLength            = sizeof(tusb_desc_device_t),
  .bDescriptorType    = TUSB_DESC_DEVICE,
  .bcdUSB             = USB_BCD,
>
>  .bDeviceClass       = TUSB_CLASS_MISC,
  .bDeviceSubClass    = MISC_SUBCLASS_COMMON,
  .bDeviceProtocol    = MISC_PROTOCOL_IAD,
>
>  .bMaxPacketSize0    = CFG_TUD_ENDPOINT0_SIZE,
  .bNumConfigurations = 0x01,
  .bReserved          = 0x00
};

And when I plug the USB into a Linux (RPi as it happens) I get an attempt to create 2 usb channels (CDC), but that fails from the host perspective...

I don't see anywhere from debugging/logging in the functionality on the embedded end where the intention is to create/define the endpoints etc. so for this reason, I'm wondering if the XUsbPs library and TinyUSB are conflicting slightly in how they should be set up -- and I'm thinking I might need to ditch the XUsbPs use (but probably copy bits in to my own implementation/port for TinyUSB)

Regards
Richard
